[JENKINS-56477] Fix to allow selecting compatible-only updates for plugins

romenrg opened the pull request:
See [JENKINS-56477](https://issues.jenkins-ci.org/browse/JENKINS-56477).

In Nov 2018 [this PR](https://github.com/jenkinsci/jenkins/pull/3715) was approved and merged, but later @daniel-beck noticed there was a bug in the selection of "compatible plugins", which was not taking all cases into account.

In this PR I have modified the condition to select the compatible plugins, so that it covers all cases and, following @oleg-nenashev suggestion, I've created a separate button to allow users to install either "All", "Compatible" or "None".

### Proposed changelog entries

* The Plugin Manager now provides improved selection for plugin updates. Users willing to update plugins can select "All", "Compatible" or "None".
* The selection of "Compatible" plugins has been fixed to exclude all that contain any warnings

### Submitter checklist

- [X] JIRA issue is well described
- [X] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)
- [X] Appropriate autotests or explanation to why this change has no tests
- [n/a] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@daniel-beck - As the original reporter of the ticket
@oleg-nenashev - Your [suggestion to provide a third option for the selection](https://github.com/jenkinsci/jenkins/pull/3715#pullrequestreview-170569822) was taken into account
@batmat - I followed your TO-DO in the UpdateSiteTest and factored out the initialization piece for the UpdateSite creation that was being used in the "minimumJavaVersion" test and is now also included in the test I added (isPluginCompatible). Notice I removed the "sites" variable, since it seemed to be unused; using just the specific site.

romenrg committed:
[JENKINS-56477] New "isCompatible" method avoiding all plugins w/ warns

Previous version was only checking "isCompatibleWithInstalledVersion", but there were several other cases that had to be checked

romenrg committed:
[JENKINS-56477] Added issue annotation to isPluginUpdateCompatible test

romenrg committed:
[JENKINS-56477] Factored out getUpdateSite, init method used by 2 tests

Also removed the "sites" list from the init, since it seemed it was not being used, only using the specific site created from the external json

romenrg committed:
[JENKINS-56477] Added a new option for plugin selection: "compatible"

Now users can select "All", "Compatible" or "None"

romenrg committed:
[JENKINS-56477] checkPluginsWithoutWarnings also unchecks now

Useful if the user first clicks "Select All" and now wants to "Select Compatible" only.

romenrg committed:
[JENKINS-56477] Added unit test for "UpdateSite.Plugin.isCompatible"

The UpdateSite class and its internal classes (e.g. Plugin) are a bit tricky to unit-test. Using some stubs to set the necessary context for the test to run

romenrg commented:
I'm attaching here a screenshot of the result of clicking the new "Compatible" link at the button of the Plugin Manager page:

<img width="1440" alt="fixed-select-compatible" src="https://user-images.githubusercontent.com/1831480/56092462-4ec26e00-5eb4-11e9-853d-d659d094e2af.png">

daniel-beck commented on core/src/main/java/hudson/model/UpdateSite.java:
```diff
@@ -1043,6 +1043,18 @@ public PluginWrapper getInstalled() {
+            return isCompatible(new PluginManager.MetadataCache());
+        }
+
+        @Exported
```
Cannot be exported due to argument, so should be throwing an exception in the remote API (or at least log a warning that this is un-exportable)

romenrg replied:
I'm not sure how the @Exported annotation is being used. I included it in both cases as I saw it in the original "isCompatibleWithInstalledVersion", but now that I'm looking to it in more detail, I assume it is only used as a way to publish the method as an API operation, right?

If that is the case, it seems that for this method overload the annotation is not necessary at all.

romenrg replied:
Removed and tested locally afterwards, working fine.

daniel-beck commented on core/src/main/java/hudson/model/UpdateSite.java:
```diff
@@ -1043,6 +1043,18 @@ public PluginWrapper getInstalled() {
             return pm.getPlugin(name);
         }
 
+        @Exported
```
Unnecessary?

romenrg replied:
I assume the fact that there is a jelly file that invokes it doesn't imply that the method has to be annotated with "Exported", right? I wasn't sure how this was being used.

If that is not a requirement, then there is no need to have it, so I'll be happy to remove it.

daniel-beck replied:
This is for the "REST" API, i.e. `/pluginManager/api` or `/updateCenter/api`, don't remember which, and entirely optional.

romenrg replied:
Thanks for the explanation, Daniel 👍

romenrg replied:
Since there have been no comments from others on this, I've applied @daniel-beck 's suggestion and removed the `@Exported` annotation. Definitely not necessary for now.

daniel-beck commented on core/src/main/resources/hudson/PluginManager/_table.js:
```diff
@@ -5,6 +5,9 @@ function checkPluginsWithoutWarnings() {
         if(candidate.type === "checkbox" && candidate.dataset.compatWarning === 'false') {
             candidate.checked = true;
         }
+        else if(inputs[i].type === "checkbox") {
```
`inputs[i]` == `candidate` so why are we not using the nicer local var here? Especially since with the next line this is really confusing code.

Also, this can probably be simplified to:

         if(candidate.type === "checkbox") {
            candidate.checked = candidate.dataset.compatWarning === 'false';
        }

romenrg replied:
You are right, should be using `candidate` instead.

That said, I really like your proposed alternative. Very elegant. Will use it :)

romenrg replied:
Done 👍

daniel-beck commented on core/src/test/java/hudson/model/PluginTest.java:
```diff
@@ -0,0 +1,42 @@
+public class PluginTest {
+
+    @Test
+    public void pluginUpdateNotCompatible() throws IOException, ExecutionException, InterruptedException {
```
`@Issue("JENKINS-56477")`

daniel-beck commented on core/src/test/java/hudson/model/PluginTest.java:
```diff
@@ -0,0 +1,42 @@
+
+import static org.hamcrest.CoreMatchers.is;
+import static org.junit.Assert.assertThat;
+import static org.mockito.Mockito.*;
```
What does the use of Mockito here gain you? Doesn't the other (modified) test not show this is unnecessary?

romenrg replied:
Yes, I've removed this unit test and I'm now leaving only the other integration test as sufficient.

romenrg committed:
[JENKINS-56477] "mock-maker-inline" incompatible w/ other tests

Had to remove the unit test checking "pluginUpdateNotCompatible. It needed "mock-maker-inline" config to mock/spy the Plugin class (which is final), but that mockito config caused conflicts with pre-existing tests.

romenrg committed:
[JENKINS-56477] Simplifying condition to check the candidate checkbox

romenrg committed:
[JENKINS-56477] Method overload with param should not be exported

daniel-beck commented on core/src/main/java/hudson/model/UpdateSite.java:
```diff
@@ -1043,6 +1043,17 @@ public PluginWrapper getInstalled() {
         }
 
+        @Exported
+        public boolean isCompatible() {
```
Would `@Restricted(NoExternalUse.class)` this, no reason IMO to expose this to plugins.

If we end up not doing that, needs Javadoc + `@since TODO`.

romenrg replied:
For this case I've included the Javadoc and the `@since TODO`.

I might be wrong, but I see this new method as the replacement of the one that was being called previously (`isCompatibleWithInstalledVersion`), which is being exported; so I thought it could be useful to export this one so that whoever uses the "old" one can switch to this new and more comprehensive check

romenrg replied:
Since nobody else has commented on this, I've applied @daniel-beck 's suggestion and made the `isCompatible` methid restricted (`@Restricted(NoExternalUse.class)`).

Definitely not necessary to expose it to plugins for now. And if in the future someone sees value in calling it from a plugin this can always be changed.

daniel-beck commented on core/src/main/java/hudson/model/UpdateSite.java:
```diff
@@ -1043,6 +1043,17 @@ public PluginWrapper getInstalled() {
+            return isCompatible(new PluginManager.MetadataCache());
+        }
+
+        public boolean isCompatible(PluginManager.MetadataCache cache) {
```
Would `@Restricted(NoExternalUse.class)` this, no reason IMO to expose this to plugins.

If we end up not doing that, needs Javadoc + `@since TODO`.

romenrg replied:
I've restricted this one, since in case some plugins want to call this, it seems they should be calling the version without the "cache" param, as is the pattern for similar methods.

daniel-beck reviewed (commented):
Minor remarks, else LGTM.

romenrg committed:
[JENKINS-56477] Javadoc for isCompatible(). Overloaded ver. restricted

Also switched the call in table.jelly to the overloaded version of isCompatible, since a "cache" var was available in the context (and was being already used by calls to the internal methods to show the warnings)

romenrg committed:
Removed empty line.

Trivial change to force re-execution of pipeline.

romenrg committed:
[JENKINS-56477] "isCompatible" set to @Restricted for now

This method is currently only being used by the Jenkins UI, so we are not being exposing it to plugins nor exporting it. If in the future that need emerges, this can be changed

romenrg commented:
In the absence of other comments, I've gone ahead and apply @daniel-beck 's suggestions, so that the new method is not exported nor exposed to plugins. In case in future someone sees value on using it from a plugin, it can always be changed.

Thanks Daniel for your comments and explanations :)

oleg-nenashev commented on core/src/main/resources/hudson/PluginManager/index.jelly:
```diff
@@ -29,7 +29,9 @@ THE SOFTWARE.
     <div style="margin-top:1em">
-      ${%Select}: <a href="javascript:checkPluginsWithoutWarnings();">${%All}</a>, <a href="javascript:toggleCheckboxes(false);">${%None}</a><br/>
+      ${%Select}: <a href="javascript:toggleCheckboxes(true);">${%All}</a>,
+                  <a href="javascript:checkPluginsWithoutWarnings();">${%Compatible}</a>,
```
I would rather put "Compatible" before "All".

oleg-nenashev reviewed (approved):
LGTM, it will definitely improve the situation.
Not sure how ATH tests would react to it tho

oleg-nenashev commented:
@daniel-beck WDYT?

oleg-nenashev commented:
On hold for this weekly to let other reviewers to provide feedback.
Will merge it if @daniel-beck or @batmat bless it

daniel-beck commented:
@oleg-nenashev 

> let other reviewers to provide feedback

This is two weeks old. People had a chance to review.

daniel-beck commented:
@oleg-nenashev 

> WDYT?

I stand by https://github.com/jenkinsci/jenkins/pull/3985#pullrequestreview-226401995. The minor remarks have since been addressed.

oleg-nenashev commented:
🚢 🇮🇹 then. Thanks @romenrg !
